### ISSUE TITLE
Remove all direct "echo"

### ISSUE THREAD

**Issue by trasher** [glpi-project/glpi#1605]
There are many "echo" directly done in the source code, this is weird.

Some methods already provide a `display` parameter that permit to either output directly with a `echo` or return a variable containing a string.
That way would also permit unit tests to be run against those methods.

This should be generalized.

Additionally, we should add a constant defining the default value for the display parameter; so we can change it quickly in the future if needed.

**Comment by remicollet** [comment]
Indeed,

And this will help to lot with PDF integration (using the same method to generate HTML for UI and PDF).

Perhaps this is also linked to have "read-only" option for form.

**Comment by trasher** [comment]
> Perhaps this is also linked to have "read-only" option for form.

Could you give me precisions?

**Comment by nbriche** [comment]
Shouldn't that be part of an effort to convert the framework to a template-based structure ?

**Comment by remicollet** [comment]
> Could you give me precisions?

In readonly mode, replace all Html::autocompletionTextField by simple output, all Dropdownn::show by getDropdownName, etc.

This will make output much lighter.

Then, we can use this small HTML code to embed it in PDF output.

> Shouldn't that be part of an effort to convert the framework to a template-based structure ?

Definitively -1 from me.
All templating engine sucks (smarty, twig, plates... they break their API too often. No value. PHP "is" a already a template engine.

**Comment by nbriche** [comment]
Without going as far as a third-party framework, I was mainly suggesting separating engine, model and presentation.  As it is, there's bits of HTML strings basically everywhere, which makes it difficult to figure out where the engine stops and the page begins, or even what the page is supposed to look like in the end.

Sure, PHP is a template engine in that it can be included inside HTML tags in master HTML files, but there's very little (if any) of that in the current code, at least that I could understand.

I know that the "separate engine, model and presentation" mantra sounds like a bad reheated trope in 2017, but honestly I would have lost my mind years ago if I didn't apply it to my own projects. Granted, I currently only have one major home-made PHP-based application (an intranet using PHPTAL); for all my other web-based works I've used CherryPy and Genshi, so maybe I'm just too used to a tiered approach.

**Comment by trasher** [comment]
@remicollet OK I see, thanks for the clarifications :)
@nbriche be able to separate contents from logic is something I have in mind; and this must be done to achieve such a thing; yes. But this also have other benefits (unit testing for example), thus this is not "officially" a first step to... nothing ;)

**Comment by trasher** [comment]
I made some tries since I've opened this one... And the conclusion is that this is not a good idea.

Direct echoes must be dropped, but the issue is more globally that the view is not separate from the logic. When I've opened this issue, I guess we should be able to make some replacements, but this is also a very huge job.

I close this one in favor of https://github.com/glpi-project/glpi/issues/4594.
